File: src/main/java/com/codingforcookies/betterrecords/RecipeRecordShuffle.java
package com.codingforcookies.betterrecords;

import com.codingforcookies.betterrecords.items.ItemURLMultiRecord;
import com.codingforcookies.betterrecords.items.ItemURLRecord;

import net.minecraft.init.Blocks;
import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;

public class RecipeRecordShuffle implements IRecipe{

	@Override
	public boolean matches(InventoryCrafting par1InventoryCrafting, World par2World){
		ItemStack record = null;
		boolean shuffle = false;
		for(int i = 0; i < par1InventoryCrafting.getSizeInventory(); ++i){
			ItemStack itemstack = par1InventoryCrafting.getStackInSlot(i);
			if(itemstack != null) {
				if(itemstack.getItem() instanceof ItemURLMultiRecord && itemstack.getTagCompound() != null) if(record != null) return false;
				else record = itemstack;
				else if(itemstack.getItem() == Item.getItemFromBlock(Blocks.redstone_torch)) if(shuffle) return false;
				else shuffle = true;
				else return false;
			}
		}
		return record != null && shuffle;
	}

	@Override
	public ItemStack getCraftingResult(InventoryCrafting par1InventoryCrafting){
		ItemStack record = null;
		boolean shuffle = false;
		for(int i = 0; i < par1InventoryCrafting.getSizeInventory(); i++){
			ItemStack itemstack = par1InventoryCrafting.getStackInSlot(i);
			if(itemstack != null) {
				if(itemstack.getItem() instanceof ItemURLRecord && itemstack.getTagCompound() != null) if(record != null) return null;
				else record = itemstack;
				else if(itemstack.getItem() == Item.getItemFromBlock(Blocks.redstone_torch)) if(shuffle) return null;
				else shuffle = true;
				else return null;
			}
		}
		if(record == null || !shuffle) return null;
		else{
			ItemStack newRecord = ItemStack.copyItemStack(record);
			if(newRecord.getTagCompound() == null) newRecord.setTagCompound(new NBTTagCompound());
			newRecord.getTagCompound().setBoolean("shuffle", true);
			return newRecord;
		}
	}

	@Override
	public int getRecipeSize(){
		return 10;
	}

	@Override
	public ItemStack getRecipeOutput(){
		return null;
	}

	@Override
	public ItemStack[] getRemainingItems(InventoryCrafting inv) {
		inv.clear();
		return new ItemStack[] {};
	}
}

File: src/main/kotlin/com/codingforcookies/betterrecords/crafting/recipe/RecipeRecordShuffle.kt
package com.codingforcookies.betterrecords.crafting.recipe

import com.codingforcookies.betterrecords.item.ItemMultiRecord
import com.codingforcookies.betterrecords.item.ItemRecord
import com.codingforcookies.betterrecords.item.ModItems
import net.minecraft.init.Blocks
import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.NonNullList
import net.minecraft.world.World
import net.minecraftforge.registries.IForgeRegistryEntry

class RecipeRecordShuffle : IForgeRegistryEntry.Impl<IRecipe>(), IRecipe {

    init {
        setRegistryName("reciperecordshuffle")
    }

    override fun canFit(width: Int, height: Int) = width * height >= 2

    override fun matches(inventoryCrafting: InventoryCrafting, worldIn: World?): Boolean {
        var foundRecord = false
        var foundTorch = false

        (0 until inventoryCrafting.sizeInventory)
                .map { inventoryCrafting.getStackInSlot(it) }
                .filter { !it.isEmpty }
                .forEach {
                    if (it.item is ItemMultiRecord && it.hasTagCompound() && !foundRecord) {
                        foundRecord = true
                    } else if (it.item == Item.getItemFromBlock(Blocks.REDSTONE_TORCH) && !foundTorch) {
                        foundTorch = true
                    } else {
                        return false
                    }
                }

        return foundRecord && foundTorch
    }

    override fun getCraftingResult(inventoryCrafting: InventoryCrafting): ItemStack? {
        val record =
                (0 until inventoryCrafting.sizeInventory)
                        .map { inventoryCrafting.getStackInSlot(it) }
                        .filter { !it.isEmpty }
                        .find { it.item is ItemRecord }

        return record?.copy()?.apply {
            if (!hasTagCompound()) {
                tagCompound = NBTTagCompound()
            }
            tagCompound!!.setBoolean("shuffle", true)
        }
    }

    override fun getRecipeOutput() = ItemStack(ModItems.itemMultiRecord)

    override fun getRemainingItems(inv: InventoryCrafting): NonNullList<ItemStack> {
        inv.clear()
        return NonNullList.create()
    }
}
